File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_MOVIMIENTOS_PREDEFINIDOS/P5-java/src/irrgarten/Game.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
import java.util.Arrays;


/** 
 *
 * @author ismael
 */

/**
 * La clase {@code Game} representa el núcleo de la lógica de un juego, que incluye 
 * a los jugadores, monstruos, un laberinto y la gestión de rondas. 
 * Controla el estado actual del juego, el jugador que está en turno, 
 * y el registro de eventos.
 */
public class Game {
    
    /**
     * Número máximo de rondas que puede tener el juego.
     */
    private static final int MAX_ROUNDS = 10;
    
    /**
     * Índice del jugador actual en el arreglo de jugadores.
     */
   private  int currentPlayerIndex;
    
    /**
     * Registro de los eventos que ocurren durante el juego.
     */
    private String log;
    
    /**
     * Arreglo de jugadores que participan en el juego.
     */
    private ArrayList<Player> players;
    
    /**
     * Jugador que está tomando el turno actual.
     */
    private Player currentPlayer;
    
    /**
     * Arreglo de monstruos que se encuentran en el juego.
     */
    private ArrayList<Monster> monsters;
    
    /**
     * Laberinto donde se desarrolla el juego.
     */
    private Labyrinth labyrinth;
    
    //------------VARIABLES PARA MI LABERINTO----------
    
    /**
     * Número de monstruos
     */
    private final int NUM_MONSTER = 3;
    
    /**
     * Número de obtáculos
     */
    private final int NUM_BLOCKS = 5;
    
    /**
     * Número de filas
     */
    private final int N_ROW = 7;
    
    /**
     * Número de columnas
     */
    private final int N_COL = 7;
    
    /**
     * Número de salida
     */
    private final int EXIT_ROW = 6;
    
    /**
     * Número de salida
     */
    private final int EXIT_COL = 6;
    
    /**
     * Número aleatorio para generar longitud del muro
     */
    private final int RANDOM_POS = 5;
    
    /**
     * Número aleatorio para generar la orientación
     */
    private final int RANDOM_ORIENTATION = 2;
    
    //------------VARIABLES PARA MI LABERINTO----------
    
    public Game(int nPlayers){
        // Inicializamos los vectores de jugadores y monstruos
        players=new ArrayList<>();
        monsters=new ArrayList<>();
        
        // Inicializar el índice del jugador actual y el log
        currentPlayerIndex = Dice.whoStarts(nPlayers);
        log = "Game has started.\n";

        // crear los jugadores y añadirlos al contenedor adecuado
        for (int i = 0; i < nPlayers; i++) {
            char c = (char) ('0' + i); //asignar índice pero en formato char
            players.add(new Player(c, Dice.randomIntelligence(), Dice.randomStrength()));
        }
        
        //asignamos el jugador actual
         currentPlayer = players.get(currentPlayerIndex);
         
        //Método que configura el laberinto
        configureLabyrinth();

        //Repartir jugadores por el laberinto
        labyrinth.spreadPlayers(players);
    }

    public boolean finished(){
        return labyrinth.haveAWinner();
    }

    /**
     * Genera una instancia de GameState integrando toda la información del
     * estado del juego.
     *
     * @return una instancia de GameState que representa el estado actual del juego.
     */
    public GameState getGameState(){
        String playerString = "";
        for (Player player : players) {
            playerString += player.toString() + "\n";
        }
        String monsterString = "";
        for (Monster monster : monsters) {
            monsterString += monster.toString() + "\n";
        }
        return new GameState(labyrinth.toString(), playerString,monsterString, currentPlayerIndex,finished(), log);
    }

    /**
     * Configura el laberinto añadiendo bloques de obstáculos y monstruos.
     * Los monstruos, además de en el laberinto, son guardados en el contenedor propio de esta clase para
     * este tipo de objetos.
     */
    public void configureLabyrinth(){
        // Crear el laberinto
        labyrinth  = new Labyrinth(N_ROW, N_COL, EXIT_ROW, EXIT_COL);
        // Añadir bloques de obstáculos
        for (int i = 0; i < NUM_BLOCKS; i++) { // Añadir 5 bloques como ejemplo
            int startRow = Dice.randomPos(labyrinth.getnRows()); // Genera una fila inicial aleatoria dentro del rango del laberinto
            int startCol = Dice.randomPos(labyrinth.getnCols()); // Genera una columna inicial aleatoria dentro del rango del laberinto
            int length = Dice.randomPos(RANDOM_POS) + 1; // Genera una longitud aleatoria para el bloque entre 1 y 5
            Orientation orientation = Dice.randomPos(RANDOM_ORIENTATION) == 0 ? Orientation.HORIZONTAL : Orientation.VERTICAL; // Determina aleatoriamente si el bloque será horizontal o vertical
            labyrinth.addBlock(orientation, startRow, startCol, length); // Añade el bloque al laberinto en la posición y orientación especificadas
        }

        // Crear los monstruos (añadimos 3 monstruos como ejemplo)
        for (int i = 0; i < NUM_MONSTER-1; i++) { // Iterar sobre cada monstruo en el arreglo de monstruos
            monsters.add(new Monster("Monster " + (i + 1), Dice.randomIntelligence(), Dice.randomStrength())); // Crear un monstruo con nombre y atributos aleatorios
        }
        monsters.add(new Monster("The Killer", 30000, 30000)); // Crear un monstruo con nombre y atributos fijos

        // Añadir monstruos
        for (Monster monster : monsters) { // Itera sobre cada monstruo en el arreglo de monstruos
            int[] pos = labyrinth.randomEmptyPos(); // Encuentra una posición vacía aleatoria en el laberinto
            labyrinth.addMonster(pos[0], pos[1], monster); // Añade el monstruo al laberinto en la posición encontrada
        }
    }

    /**
     * Actualiza los dos atributos que indican el jugador (current*) con el turno pasando
     * al siguiente jugador.
     */
    public void nextPlayer(){
        currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        currentPlayer = players.get(currentPlayerIndex);
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el jugador ha ganado el combate. También añade el indicador de 
     * nueva línea al final.
     */
    public void logPlayerWon(){
        log += "Player " + currentPlayer.getNumber() + " won the combat.\n";
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el monstruo ha ganado el combate. También añade el indicador de 
     * nueva línea al final.
     */
    public void logMonsterWon(){
        log += "Monster won the combat against player " + currentPlayer.getNumber() + ".\n";
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el jugador ha resucitado. También añade el indicador de nueva línea * al final.
     */
    public void logResurrected(){
        log += "Player " + currentPlayer.getNumber() + " was resurrected.\n";
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el jugador ha perdido el turno por estar muerto. También añade el indicador de nueva línea al final.
     */
    public void logPlayerSkipTurn() {
        log += "Player " + currentPlayer.getNumber() + " skipped the turn because they are dead.\n";
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el jugador no ha seguido las instrucciones del jugador humano (no fue posible).
     * También añade el indicador de nueva línea al final.
     */
    public void logPlayerNoOrders() {
        log += "Player " + currentPlayer.getNumber() + " did not follow the human player's instructions.\n";
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el mensaje
     * indicando que el jugador se ha movido a una celda vacía o no le ha sido posible moverse.
     * También añade el indicador de nueva línea al final.
     */
    public void logNoMonster() {
        log += "Player " + currentPlayer.getNumber() + " moved to an empty cell or couldn't move.\n";
        //y cuando no se puede mover que?
    }

    /**
     * Añade al final del atributo log (concatena cadena al final) el
     * mensaje que se han producido rounds de max rondas de combate. 
     * También añade el indicador de
     * nueva línea al final.
     */
    public void logRounds(int rounds, int max) {
        log += "Rounds played: " + rounds + " out of " + max + ".\n";
    }

    /**
     * Ejecuta el siguiente paso en el juego para el jugador actual.
     * 
     * @param preferredDirection La dirección preferida en la que el jugador quiere moverse.
     * @return true si el juego ha terminado, false en caso contrario.
     */
    public boolean nextStep(Directions preferredDirection) {
        log="";
        boolean dead = currentPlayer.dead();
        
        if(!dead){
            // Se obtiene la dirección real a la que se mueve el jugador.
            // Si no se puede mover, se indica en el log
            Directions direction = actualDirection(preferredDirection);
            System.out.print("direction es: " + direction.toString() + "y preferred es" + preferredDirection.toString() );
            if(direction != preferredDirection) logPlayerNoOrders();
        
            // Se mueve al jugador y se obtiene el monstruo que haya en la casilla
            Monster monster = labyrinth.putPlayer(preferredDirection, currentPlayer);

            // Se estudia si hay un monstruo en la casilla a la que se ha movido el jugador
            if(monster == null){
                logNoMonster();
            }
            else{
                //comienza el combate
                GameCharacter winner = combat(monster);
                //recompensa
                manageReward(winner);
            }
        }
        else{
            manageResurrection();
        }
        boolean endGame = finished();
        if(!endGame) nextPlayer(); //comprobamos si ha finalizado el juego, si no pasamos al siguiente jugador
        
        return endGame;
    }

    /**
     * Determina la dirección real en la que el jugador puede moverse en función de su dirección preferida
     * y los movimientos válidos disponibles desde su posición actual en el laberinto.
     *
     * @param preferredDirection La dirección en la que el jugador prefiere moverse.
     * @return La dirección real en la que el jugador se moverá, que puede ser la dirección preferida
     *         si es válida, o otra dirección válida si la preferida no es posible.
     */
    public Directions actualDirection(Directions preferredDirection) {
        // Se obtiene la fila y columna actual del jugador
        int currentRow=currentPlayer.getRow();
        int currentCol=currentPlayer.getCol();
        // Direcciones posibles desde la casilla actual
        ArrayList<Directions> validMoves = labyrinth.validMoves(currentRow, currentCol);
        // Se mueve al jugador y devolvemos esa dirección
         return currentPlayer.move(preferredDirection, validMoves);
    }

    /**
     * Simula un combate entre el jugador actual y un monstruo.
     * El combate consiste en una serie de rondas de ataque y defensa hasta que uno de los participantes pierda o se alcance el número máximo de rondas.
     *
     * @param monster el monstruo contra el que luchar
     * @return el ganador del combate, ya sea GameCharacter.PLAYER o GameCharacter.MONSTER
     */
    public GameCharacter combat(Monster monster) {
        int rounds = 0; //inicializamos el numero de rondas a 0
        
        // Suponemos que el jugador ganará, y empieza este atacando.
        GameCharacter winner = GameCharacter.PLAYER;
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        
        // Bucle que simula el combate entre el jugador y el monstruo
        while(!lose && rounds<MAX_ROUNDS){
            rounds++; // incrementamos rondas
            
            // Suponemos que el monstruo ganará, y continúa este atacando.
            winner = GameCharacter.MONSTER; 
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            
            if(!lose){ // si no ha perdido, suponemos que el jugador gana y este seguirá atacando
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        logRounds(rounds,MAX_ROUNDS);
        
        return winner;
    }

    /**
     * Gestiona la distribución de recompensas según el personaje del juego que ganó.
     * Si el jugador gana, recibe una recompensa y se crea una entrada en el registro para la victoria del jugador.
     * Si el monstruo gana, se crea una entrada en el registro para la victoria del monstruo.
     *
     * @param winner el personaje del juego que ganó, ya sea GameCharacter.PLAYER u otro personaje.
     */
    public void manageReward(GameCharacter winner) {
        if(winner==GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            logPlayerWon();
        }
        else{
            logMonsterWon();
        }
        
    }

    /**
     * Gestiona el proceso de resurrección para el jugador actual.
     * Si el jugador es resucitado, llama al método resurrect en el jugador actual
     * y registra el evento de resurrección. Si el jugador no es resucitado, 
     * registra que el jugador perderá su turno.
     */
    public void manageResurrection() {
        boolean resurrect = Dice.resurrectPlayer();
        if(resurrect){
            currentPlayer.resurrect();
            logResurrected();
            
            //añadir la funcionalidad del fuzzyplayer
            FuzzyPlayer fuzzyPlayer = new FuzzyPlayer(currentPlayer);
            players.set(currentPlayerIndex, fuzzyPlayer);
            labyrinth.PlayerTOfuzzyPlayer(fuzzyPlayer);
        }
        else{
            logPlayerSkipTurn();
        }
    }

    //------FUNCIÓN EXTRA------
    /**
     * Función que me pinta el laberinto que tengo de la mejor manera posible
     */
    /*public void printLabyrinth(){
        System.out.println("\n Laberinto: \n");
        for(int i = 0; i <labyrinth.getnRows(); i++){
            for(int j = 0; j <labyrinth.getnCols(); j++){
                System.out.print("[ "+labyrinth.getLabyrinth(i, j)+" ]  ");
            }
            System.out.print("\n");
        }
    }*/

}

File: Asignaturas/Tercer Año/PDOO/Examenes/SolucionParcialPracticas2/P5-Ruby/lib/irrgarten/game.rb
#encoding:utf-8
require_relative 'dice'
require_relative 'player'
require_relative 'monster'
require_relative 'labyrinth'
require_relative 'game_state'
require_relative 'game_character'
require_relative 'orientation'
require_relative 'fuzzy_player'
require_relative 'super_player'



# Módulo principal del juego Irrgarten, que contiene las clases de jugadores, monstruos y el laberinto.
module Irrgarten 
    # Representa el juego, manejando el estado actual, los jugadores, monstruos y el laberinto.
    class Game
        # Número máximo de rondas en el juego.
        @@MAX_ROUNDS = 10

        #--------VARIABLES PARA MI LABERINTO----------
        # Número de monstruos
        @@NUM_MONSTER = 3

        # Número de obstáculos
        @@NUM_BLOCKS = 5

        # Número de filas
        @@N_ROW = 7

        # Número de columnas
        @@N_COL = 7

        # Número de salida (fila)
        @@EXIT_ROW = 6

        # Número de salida (columna)
        @@EXIT_COL = 6

        # Número aleatorio para generar longitud de muro
        @@RANDOM_POS = 5

        # Número aleatorio para generar la orientación
        @@RANDOM_ORIENTATION = 2

        #------Accessors para las variables de clase------

        # @!attribute [rw] currentPlayerIndex
        #   @return [Integer] Índice del jugador actual en la lista de jugadores.
        attr_accessor :currentPlayerIndex

        # @!attribute [rw] log
        #   @return [Array<String>] Registro de eventos ocurridos durante el juego.
        attr_accessor :log

        # @!attribute [rw] players
        #   @return [Array<Player>] Lista de jugadores en el juego.
        attr_accessor :players

        # @!attribute [rw] currentPlayer
        #   @return [Player] El jugador que actualmente tiene el turno.
        attr_accessor :currentPlayer

        # @!attribute [rw] monsters
        #   @return [Array<Monster>] Lista de monstruos en el juego.
        attr_accessor :monsters

        # @!attribute [rw] lab
        #   @return [Labyrinth] El laberinto en el que se juega.
        attr_accessor :labyrinth

        #------Accessors para las variables de clase------

        # Inicializa una nueva instancia del juego con los jugadores, monstruos y el laberinto configurados.
        #
        # @param nPlayers [Integer] Número de jugadores en el juego.
        def initialize(nPlayers)
            @currentPlayerIndex = Dice.whoStarts(nPlayers)

            @log = ""

            @monsters = Array.new(@@NUM_MONSTER) #ya que los decidimos nosotros, inicializo ya el constructor aquí
            @players = Array.new(nPlayers)

            nPlayers.times do |i|
                @players[i]= Player.new(i,Dice.randomIntelligence, Dice.randomStrength)
            end

            @currentPlayer = @players[@currentPlayerIndex]

            configureLabyrinth()

            @labyrinth.spreadPlayers(players)
        end

        # Configura el laberinto añadiendo bloques y colocando monstruos.
        #
        # @return [void]
        def configureLabyrinth 
            #creamos el laberinto
            @labyrinth = Labyrinth.new(@@N_ROW,@@N_COL,@@EXIT_ROW,@@EXIT_COL) 

            #añiadimos 5 bloques por ejemplo
            @@NUM_BLOCKS.times do
                startRow = Dice.randomPos(labyrinth.nRows)
                startCol = Dice.randomPos(labyrinth.nCols)
                length = Dice.randomPos(@@NUM_BLOCKS) # definimos la longitud de los blocks
                orientation = Dice.randomPos(@@RANDOM_POS) == 0 ? Orientation::HORIZONTAL : Orientation::VERTICAL
                labyrinth.addBlock(orientation, startRow, startCol, length)
            end

            #creamos los monstruos
            monsters[0] = Monster.new("Goblin", Dice.randomIntelligence, Dice.randomStrength);
            monsters[1] = Monster.new("Orc", Dice.randomIntelligence, Dice.randomStrength);
            monsters[2] = Monster.new("Troll", Dice.randomIntelligence, Dice.randomStrength);
            monsters[2] = Monster.new("THEKILLER", 100000, 100000); #para probar que el jugador no puede ganar


            #añadimos los monstruos
            monsters.each do |monster|
                pos = @labyrinth.randomEmptyPos()
                @labyrinth.addMonster(pos[0], pos[1], monster)
            end
        end



        # Verifica si el juego ha terminado.
        #
        # @return [Boolean] `true` si el juego ha finalizado, `false` en caso contrario.
        def finished
            @labyrinth.haveAWinner
        end

        # Obtiene el estado actual del juego.
        #
        # @return [GameState] Una nueva instancia de GameState que representa el estado actual del juego.
        def getGameState
            player_string = ""
            @players.each do |player|
                player_string += player.to_s + "\n"
            end
    
            monster_string = ""
            @monsters.each do |monster|
                monster_string += monster.to_s + "\n"
            end
    
            GameState.new(@labyrinth.to_s, player_string, monster_string, @currentPlayerIndex, finished, @log)
        end

        

        # Cambia el turno al siguiente jugador.
        #
        # @return [void]
        def nextPlayer
            @currentPlayerIndex = (@currentPlayerIndex + 1) % @players.size
            @currentPlayer = @players[@currentPlayerIndex]
        end

        # Registra que el jugador actual ganó el combate en el registro de eventos.
        #
        # @return [void]
        def logPlayerWon
            
            @log += "Player " + @currentPlayer.number.to_s + " won the combat.\n"
        end

        # Registra que el monstruo ganó el combate contra el jugador actual en el registro de eventos.
        #
        # @return [void]
        def logMonsterWon
            @log += "Monster won the combat against player " + currentPlayer.number.to_s + ".\n"
        end

        # Registra en el log que el jugador actual ha sido resucitado.
        #
        # @return [void]
        def logResurrected
            @log += "Player " + currentPlayer.number.to_s + " was resurrected.\n"
        end

        # Registra en el log que el jugador actual se ha saltado el turno por estar muerto.
        #
        # @return [void]
        def logPlayerSkipTurn
            
            @log += "Player " + currentPlayer.number.to_s + " skipped the turn because they are dead.\n"
        end

        # Registra en el log que el jugador actual no siguió las instrucciones del jugador humano.
        #
        # @return [void]
        def logPlayerNoOrders
            
            @log += "Player " + currentPlayer.number.to_s + " did not follow the human player's instructions.\n"
        end

        # Registra en el log que el jugador actual se movió a una celda vacía o no pudo moverse.
        #
        # @return [void]
        def logNoMonster
            
            @log += "Player " + currentPlayer.number.to_s + " moved to an empty cell or could not move.\n"
        end

        # Registra en el log la ronda actual y el número máximo de rondas.
        #
        # @param rounds [Integer] Número de rondas transcurridas.
        # @param max [Integer] Número máximo de rondas.
        # @return [void]
        def logRounds(rounds, max)
            @log += "Rounds played " + rounds.to_s + " out of " + max.to_s + ".\n"
        end


        # Ejecuta el siguiente paso en el juego para el jugador actual.
        #
        # @param preferredDirection [Symbol] La dirección preferida en la que el jugador quiere moverse.
        # @return [Boolean] Devuelve true si el juego ha terminado, de lo contrario false.
        #
        # El método realiza las siguientes acciones:
        # - Verifica si el jugador actual está muerto. Si es así, maneja el proceso de resurrección.
        # - Determina la dirección real en la que el jugador se moverá.
        # - Registra un mensaje si la dirección real difiere de la dirección preferida.
        # - Intenta mover al jugador en el laberinto.
        # - Si se encuentra con un monstruo, se inicia un combate y se recompensa al ganador.
        # - Verifica si el juego ha terminado.
        # - Si el juego no ha terminado, procede al siguiente jugador.
        def nextStep(preferredDirection)
            @log=""
            dead=@currentPlayer.dead
            unless dead
                direction = actualDirection(preferredDirection)

                if direction != preferredDirection
                    logPlayerNoOrders
                end
                monster = labyrinth.putPlayer(preferredDirection, @currentPlayer)

                if (monster == nil)
                    logNoMonster
                else 
                    winner = combat(monster)
                    manageReward(winner)
                end
            else
                manageResurrection
            end
            endGame = finished
            unless endGame
                nextPlayer
            end
            endGame
        end

    
        
        # Determina la dirección real en la que el jugador puede moverse según su dirección preferida y los movimientos válidos.
        #
        # @param preferredDirection [Symbol] la dirección en la que el jugador prefiere moverse (:up, :down, :left, :right)
        # @return [Boolean] true si el jugador se mueve con éxito en la dirección preferida, false en caso contrario
        def actualDirection(preferredDirection)
            currentRow = @currentPlayer.row
            currentCol = @currentPlayer.col
            validMoves = @labyrinth.validMoves(currentRow, currentCol)
            @currentPlayer.move(preferredDirection, validMoves)
        end

        # Inicia una secuencia de combate entre el jugador actual y un monstruo.
        #
        # @param monster [Monster] El monstruo contra el que el jugador luchará.
        # @return [GameCharacter] El ganador del combate, ya sea GameCharacter::PLAYER o GameCharacter::MONSTER.
        #
        # El combate consiste en una serie de rondas donde el jugador y el monstruo se turnan para atacarse.
        # El combate continúa hasta que se alcanza el número máximo de rondas o uno de los participantes pierde toda su salud.
        # La función registra el número de rondas que tuvieron lugar y devuelve el ganador del combate.
        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER

            playerAttack = currentPlayer.attack
            lose = monster.defend(playerAttack)
            
            while !lose && rounds < @@MAX_ROUNDS
                winner = GameCharacter::MONSTER
                rounds += 1
                monster_attack = monster.attack
                lose = currentPlayer.defend(monster_attack)
            
                unless lose
                    player_attack = currentPlayer.attack
                    winner = GameCharacter::PLAYER
                    lose = monster.defend(player_attack)
                end
            end
            logRounds(rounds, @@MAX_ROUNDS)
            
            winner

        end

        # Maneja la recompensa para el ganador del combate.
        #
        # @param winner [GameCharacter] El ganador del combate.
        # @return [void]
        def manageReward(winner)
            if winner == GameCharacter::PLAYER
                currentPlayer.receiveReward
                logPlayerWon
            else
                logMonsterWon
            end
        end

        # Maneja la resurrección del jugador actual.
        #
        # @return [void]
        def manageResurrection
            resurrect = Dice.resurrectPlayer
            if resurrect
                currentPlayer.resurrect
                logResurrected
                
                p = Dice.WhoPlayerIs
                if(p == PlayerTypes::FUZZY)
                  fuzzy = FuzzyPlayer.new(@currentPlayer)
                  @players[@currentPlayerIndex] = fuzzy
                  @labyrinth.PlayerTOFuzzyPlayer(fuzzy)
                else
                  superPlayer = SuperPlayer.new(@currentPlayer)
                  @players[@currentPlayerIndex] = superPlayer
                  @labyrinth.PlayerTOSuperPlayer(superPlayer)
                end
                
            else
                logPlayerSkipTurn
            end
        end

    end
end


